writer.go

```diff
@@ -992,6 +859,301 @@ func (w *Writer) chooseTopic(msg Message) (string, error) {
 	return w.Topic, nil
 }
 
+type batchQueue struct {
+	queue []*writeBatch
+
+	cond   sync.Cond
+	closed bool
+}
+
+func (b *batchQueue) Put(batch *writeBatch) bool {
+	b.cond.L.Lock()
+	defer b.cond.L.Unlock()
+	defer b.cond.Broadcast()
+
+	if b.closed {
+		return false
+	}
+	b.queue = append(b.queue, batch)
+	return true
+}
+
+func (b *batchQueue) Get(batches []*writeBatch) []*writeBatch {
+	b.cond.L.Lock()
+	defer b.cond.L.Unlock()
+
+	for len(b.queue) == 0 && !b.closed {
+		b.cond.Wait()
+	}
+
+	var i int
+	for i = 0; i < cap(batches) && i < len(b.queue); i++ {
+		batches = append(batches, b.queue[i])
+		b.queue[i] = nil
+	}
+
+	b.queue = b.queue[i:]
+
+	return batches
+}
+
+func (b *batchQueue) Close() {
+	b.cond.L.Lock()
+	defer b.cond.L.Unlock()
+	defer b.cond.Broadcast()
+
+	b.closed = true
+}
+
+func newBatchQueue(initialSize int) *batchQueue {
+	return &batchQueue{
+		queue: make([]*writeBatch, 0, initialSize),
+		cond: sync.Cond{
+			L: &sync.Mutex{},
+		},
+	}
+}
+
+// ptWriter is a writer for a topic-partion pair. It maintains messaging order
+// across batches of messages.
+type ptWriter struct {
```

Comment:
I was torn about including topic in the name since these writers are tied to a particular topic-partition pair but I prefer `partitionWriter` as well.